src/network/http/websocket.cpp

```diff
@@ -188,7 +165,30 @@ namespace fc { namespace http {
               return _ws_connection->get_request_header(key);
             }
 
+            /****
+             * @brief retrieves the remote hostname
+             * 
+             * @param forward_header_key the key to look at in the request header
+             * @returns the value in the header, otherwise the remote endpoint
+             */
+            virtual std::string get_remote_ip(const std::string& forward_header_key) override
+            {
+               if (last_ip.empty() || last_forward_header_key != forward_header_key)
+               {
+                  // refresh the cache
+                  last_forward_header_key = forward_header_key;
+                  if (!forward_header_key.empty())
+                     last_ip = get_request_header(forward_header_key);
+                  if (last_ip.empty())
+                     last_ip = _ws_connection->get_remote_endpoint();
+               }
+               return last_ip;
+            }
+
             T _ws_connection;
+            // cache the value of the remote IP
+            std::string last_forward_header_key;
```

Comment:
I didn't get why we need this. Will the key change after a while? Will the header change after the connection is established? Will the connection be reopened after closed? Will the object be reused for a new connection?

Reply:
I agree this is not necessary. The header key should be set in the websocket_server before any connections are established, and it shouldn't change after that, and even if it *is* changed the change need not apply to new connections. KISS, IMO.

Reply:
Do we need to handle HTTP connections with `connection: keep-alive` header?

Reply:
No. http_handler always closes the connection after sending a response, and for a connection that has been upgraded to websocket protocol there will be no more http requests either.

Reply:
So it means the `connection: keep-alive` header is ignored. Are we going to support it in the future?

Reply:
If I understand this correctly, the http_handler is called for every http *request*. So we create and close a `websocket_connection` for each request, not for each actual connection. (This also means I was wrong above - we don't close the *http* connection after sending the response, but only our own *websocket_connection*.)

So even in the presence of a keep-alive http *connection*, the code in its current form should retrieve the remote address separately for each *request*.

Reply:
According to [page 21 of RFC 6455](https://tools.ietf.org/html/rfc6455#page-21), to establish a websocket connection, the request must include a `connection` header with the `upgrade` keyword inside, however, the RFC doesn't mention "keep-alive", so I think it's implementation-dependent.
```
The client's opening handshake consists of the following parts.
...
   3.   An |Upgrade| header field containing the value "websocket",
        treated as an ASCII case-insensitive value.

   4.   A |Connection| header field that includes the token "Upgrade",
        treated as an ASCII case-insensitive value.
...
```

When the connection is upgraded to a websocket connection, there is only one set of HTTP request headers attached to the connection, all following request messages must have the same original IP address (neither IP address nor forward-key would change), so it's best to get the address in `set_open_handler()` and use it in `set_message_handler()`, but not put all code in `set_message_handler()`.

When the connection is a pure HTTP connection, there may be several requests received in the connection, we may probably need to handle `connection: keep-alive` header. If a new websocket_connection object is created for each request, then the header can be ignored. Anyway, we want to extract real IP addresses of the clients and log them, so we need to update `set_http_handler()`.